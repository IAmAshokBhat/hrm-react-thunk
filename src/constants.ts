export interface ILeave {
  leaveType: string;
  from: Date;
  to: Date;
  status: string;
  noOfDays: number;
}

export interface ILeaveInfo {
  from: string;
  to: string;
  category: string;
}

export interface ISelectLeaveType {
  value: number;
  label: string;
  balance: number;
}

export enum ACTIONS {
  GET_USERS = 'GET_USERS',
  GET_USER_DETAILS = 'GET_USER_DETAILS',
  GET_ATTENDANCE_DETAILS = 'GET_ATTENDANCE_DETAILS',
  GET_LEAVE_BALANCE = 'GET_LEAVE_BALANCE',
  GET_EVENT_DETAILS = 'GET_EVENT_DETAILS',
  GET_ALL_LEAVES = 'GET_ALL_LEAVES',
  APPLY_LEAVE = 'APPLY_LEAVE',
  ADD_ATTENDANCE = 'ADD_ATTENDANCE',
  GET_LOGIN = 'GET_LOGIN',
  LOGIN_LOADING = 'GET_LOGIN_LOADING',
  LOGIN_SUCCESS = 'GET_LOGIN_SUCCESS',
  LOGIN_FAILURE = 'GET_LOGIN_FAILURE',
  GET_DESIGNATION_LIST = 'GET_DESIGNATION_LIST',
  REGISTER = 'REGISTER'
}

export interface IUserAPI {
  user_id: number;
  name: string;
}
export interface IUserListAPI {
  data: IUserAPI[];
}

export interface IUserInfoDataAPI {
  user_id: number;
  name: string;
  designation: string;
  manager: string;
  email: string;
}

export interface IUserInfoAPI {
  data: IUserInfoDataAPI[];
}

export interface IUserInfo {
  userId: number;
  name: string;
  designation: string;
  manager: string;
  email: string;
}

interface ILeaveBalanceItemAPI {
  leave_type_id: number;
  leave_type: string;
  balance: number;
}

export interface ILeaveBalanceAPI {
  data: ILeaveBalanceItemAPI[];
}

export interface ILeaveBalance {
  leaveTypeId: number;
  leaveType: string;
  balance: number;
}

interface IAttendanceItemAPI {
  name: string;
  attendance_date: string;
  start_time: string;
  end_time: string;
}

export interface IAttendanceDetailsAPI {
  data: IAttendanceItemAPI[];
}

export interface IAttendanceItem {
  name: string;
  attendanceDate: string;
  startTime: string;
  endTime: string;
  noOfHours: number;
}

export interface IError {
  ok: boolean;
  message: string;
  status: number;
}
// eslint-disable-next-line @typescript-eslint/no-explicit-any
interface IAsyncState<T = any> {
  error: IError | null;
  loading: boolean;
  value: T;
}

export interface IDefaultRoot {
  userList: IAsyncState<ILabelValue[]>;
  userInfo: IAsyncState<IUserInfo | null>;
  leaveBalance: IAsyncState<ILeaveBalance[]>;
  attendanceDetails: IAsyncState<IAttendanceItem[]>;
  evetDetails: IAsyncState<IEvent[]>;
  allLeaves: IAsyncState<ILeave[]>;
  leaveCountByMonth: IAsyncState<ILeaveCountByMonth[] | []>;
  applyLeave: IAsyncState<IGenericResponse>;
  addAttendance: IAsyncState<IGenericResponse>;
  loginDetails: IAsyncState<ILoginAPI>;
  designationList: IAsyncState<ILabelValue[]>;
  registeration: IAsyncState<IGenericResponse>;
}

export interface IAction {
  type: string;
}

export interface IUserListAction extends IAction {
  payload: IUserAPI[] | null;
}

export interface IUserInfoAction extends IAction {
  payload: IUserInfoDataAPI[] | null;
}

export interface ILeaveBalanceAction extends IAction {
  payload: ILeaveBalanceItemAPI[] | null;
}

export interface IAttendanceDetailsAction extends IAction {
  payload: IAttendanceItemAPI[] | null;
}

export interface IEventDetailsAction extends IAction {
  payload: IEventItemAPI[] | null;
}

interface ILeaveItemAPI {
  from_date: string;
  to_date: string;
  leave_type: string;
  status: number;
}

export interface IAllLeavesAction extends IAction {
  payload: ILeaveItemAPI[] | null;
}

export interface IAllLeavesAPI {
  data: ILeaveItemAPI[];
}

//TODO
export type TFAttendanceDetailsAction = (
  id: number,
  from: string,
  to: string
) => Promise<Response>;

interface IEventItemAPI {
  attendance_date?: string;
  start_time?: string;
  end_time?: string;
  from_date?: string;
  to_date?: string;
  leave_type?: string;
}

export interface IEventDetailsAPI {
  data: IEventItemAPI[];
}

export const DATE_FORMAT = 'DD-MM-YYYY';
export const DATE_FORMAT_CALENDAR = 'YYYY-MM-DD';

export enum EventType {
  WORKING = 'WORKING',
  LEAVE = 'LEAVE'
}

export interface IEvent {
  title: string;
  start: string;
  end: string;
  allDay?: boolean;
}

export interface IKeyValue {
  [key: string]: number;
}

export interface ILeaveCountByMonth {
  days: number;
  month: string;
}

export interface INewLeave {
  leaveTypeId: number;
  fromDate: string;
  toDate: string;
  userId: number;
}

export interface IGenericResponse<T = unknown> {
  status: number;
  message: string;
  data: T;
}

export interface IGenericAction extends IAction {
  payload: IGenericResponse | null;
}

export const enum NotificationType {
  SUCCESS = 'success',
  ERROR = 'error',
  WARNING = 'warning',
  INFO = 'info'
}

export interface INewAttendance {
  userId: number;
  date: string;
  inTime: string;
  outTime: string;
}

export interface ILoginData {
  userId: number;
  password: string;
}

export interface ILoginAPI {
  token: string;
  issued: number;
  expires: number;
  loginStatus: number;
  message: string;
  userId: number;
}

export interface ILoginAction extends IAction {
  payload: ILoginAPI | null;
}

export interface IRegistration {
  name: string;
  password: string;
  confirmPassword: string;
  email: string;
}

export interface ILabelValue {
  value: number;
  label: string;
}

export interface IDesignationListAPI {
  data: ILabelValue[];
}

export interface IDesignationListAction extends IAction {
  payload: ILabelValue[] | null;
}

export interface IRegisterAPI {
  name: string;
  designation: number;
  manager: number;
  password: string;
  email: string;
}

export interface IRegisterAction extends IAction {
  payload: IRegisterAPI | null;
}

export type ExpirationStatus = 'expired' | 'active' | 'grace';

export interface Session {
  id: number;
  dateCreated: number;
  username: string;
  issued: number;
  expires: number;
}

export type DecodeResult =
  | {
      type: 'valid';
      session: Session;
    }
  | {
      type: 'integrity-error';
    }
  | {
      type: 'invalid-token';
    };

export type TAnyObject = { [key: string]: any }; // eslint-disable-line @typescript-eslint/no-explicit-any

export type TAny = any; // eslint-disable-line @typescript-eslint/no-explicit-any

export const defaultAction = { type: '', payload: null };

export const API_URL = process.env.REACT_APP_API_URL;

export const defaultHeaders = { headers: { 'X-JWT-Token': undefined } };
