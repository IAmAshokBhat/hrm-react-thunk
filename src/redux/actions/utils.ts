import { Dispatch } from 'redux';
import { TFPromise, TAnyObject } from '../../contracts';

const setLoadingAction = (actionName: string) => ({
  type: `${actionName}_LOADING`
});

const fetchSuccessAction = (
  actionName: string,
  payload: TAnyObject,
  reqCmpltPayload: boolean = false
) => ({
  type: `${actionName}_SUCCESS`,
  payload: reqCmpltPayload ? payload : payload.data
});

export const generateAction = (
  actionName: string,
  dispatch: Dispatch,
  api: TFPromise,
  payload?: TAnyObject,
  reqCmpltPayload?: boolean
) => {
  dispatch(setLoadingAction(actionName));
  return api(payload)
    .then((response) => response.json())
    .then((json) => {
      dispatch(fetchSuccessAction(actionName, json, reqCmpltPayload));
    });
};
