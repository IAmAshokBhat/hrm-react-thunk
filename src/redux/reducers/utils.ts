import { TFGenerateReducer } from '../../contracts';

export const generateReducer: TFGenerateReducer = (
  state,
  action,
  actionName
) => {
  switch (action.type) {
    case `${actionName}_LOADING`:
      return { ...state, loading: true };
    case `${actionName}_SUCCESS`:
      return { ...state, loading: false, value: action.payload || [] };
    case `${actionName}_FAILURE`:
      return { error: action.payload };
    default:
      return state;
  }
};
