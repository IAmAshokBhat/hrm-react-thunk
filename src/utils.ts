export const isValidToken = () => {
  let tokenExpiry = Number(localStorage.getItem('expires'));
  return new Date(tokenExpiry) > new Date();
};

export const getToken = () => localStorage.getItem('userToken') || '';

export const getUserId = () => Number(localStorage.getItem('userId')) || -1;

interface IRequestOptions {
  method?: string;
  headers?: HeadersInit;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  body?: any;
}

const customHeaders = {
  'X-JWT-Token': getToken(),
  'Content-Type': 'application/json'
};

// const isTokenNearExpiry = () => {
//   if (localStorage.getItem('expires') === null) {
//     return false;
//   }
//   const tokenExpiry = Number(localStorage.getItem('expires'));

//   const minutesToExpiry = moment(tokenExpiry).diff(new Date(), 'minutes');

//   return minutesToExpiry <= 15;
// };

export const fetchWithToken = (url: string, options?: IRequestOptions) => {
  const requestOptions: IRequestOptions = {
    headers: { ...customHeaders }
  };
  if (options) {
    const { method = 'GET', headers, body } = options;
    requestOptions.body = body;
    requestOptions.method = method;
    requestOptions.headers = { ...requestOptions.headers, ...headers };
  }
  //   if (isTokenNearExpiry()) {
  //     console.log('Token near expiry');
  //     reFetchToken().then(() => {
  //       return fetch(url, requestOptions);
  //     });
  //   }
  return fetch(url, requestOptions);
};

// const reFetchToken = () => {
//   return new Promise(async (res, rej) => {
//     await fetch('/refresh', { headers: customHeaders })
//       .then((response) => response.json())
//       .then((json) => {
//         localStorage.setItem('userToken', json.token);
//         localStorage.setItem('expires', json.expires);
//         res(json);
//       });
//   });
// };
