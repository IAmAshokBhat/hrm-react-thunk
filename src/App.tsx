import { Header } from './components/core/AppBar';
import { Outlet } from 'react-router-dom';
import { Footer } from './components/core/Footer';
import SpeedDial from '@mui/material/SpeedDial';
import SpeedDialIcon from '@mui/material/SpeedDialIcon';
import SpeedDialAction from '@mui/material/SpeedDialAction';
import { useState } from 'react';
import MoreTimeIcon from '@mui/icons-material/MoreTime';
import EventBusyIcon from '@mui/icons-material/EventBusy';
import { AddAttendance } from './components/AddAttendance';
import { ApplyLeave } from './components/ApplyLeave';
import { isValidToken } from './utils';

function App() {
  const [open, setOpen] = useState(false);
  const [openAttendanceDialog, setOpenAttendanceDialog] = useState(false);
  const [openLeaveDialog, setOpenLeaveDialog] = useState(false);

  const handleOpen = () => setOpen(true);

  const handleClose = () => setOpen(false);

  const handleAddAttendance = () => setOpenAttendanceDialog(true);

  const handleApplyLeave = () => setOpenLeaveDialog(true);

  const actions = [
    {
      icon: <MoreTimeIcon />,
      name: 'Add Attendance',
      onAction: handleAddAttendance
    },
    {
      icon: <EventBusyIcon />,
      name: 'Apply Leave',
      onAction: handleApplyLeave
    }
  ];

  return (
    <div className="App">
      <Header />

      <body style={{ margin: ' 20px auto', maxWidth: '1536px' }}>
        <Outlet />
        {isValidToken() && (
          <>
            <SpeedDial
              ariaLabel="All Actions Of Leave Management System"
              sx={{ position: 'absolute', bottom: 35, right: 128 }}
              icon={<SpeedDialIcon />}
              onClose={handleClose}
              onOpen={handleOpen}
              open={open}
            >
              {actions.map(({ name, icon, onAction }) => (
                <SpeedDialAction
                  key={name}
                  icon={icon}
                  tooltipTitle={name}
                  onClick={onAction}
                />
              ))}
            </SpeedDial>
            <AddAttendance
              open={openAttendanceDialog}
              onClose={() => setOpenAttendanceDialog(false)}
            />
            <ApplyLeave
              open={openLeaveDialog}
              onClose={() => setOpenLeaveDialog(false)}
            />
          </>
        )}
      </body>

      <Footer />
    </div>
  );
}

export default App;
