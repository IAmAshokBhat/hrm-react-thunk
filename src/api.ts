import {
  ILoginData,
  INewAttendance,
  INewLeave,
  IRegisterAPI
} from './constants';
import { IDetailsAPIReq, IDetailsAPIReqWithDates } from './contracts';
import { fetchWithToken, getToken } from './utils';

const headers = {
  'X-JWT-Token': getToken()
};
const API_URL = process.env.REACT_APP_API_URL;

export const getUserListAPI = () => fetchWithToken('user/list');

export const getUserDetailsAPI = ({ id }: IDetailsAPIReq) =>
  fetchWithToken(`${API_URL}api/users/${id}`);

export const getAttendanceDetailsAPI = ({
  id,
  from,
  to
}: IDetailsAPIReqWithDates) =>
  fetchWithToken(`${API_URL}api/attendance/${id}?from=${from}&to=${to}`);

export const getLeaveBalanceAPI = ({ id }: IDetailsAPIReq) =>
  fetchWithToken(`${API_URL}api/leaves/${id}`);

export const getEventDetailsAPI = ({ id, from, to }: IDetailsAPIReqWithDates) =>
  fetchWithToken(`${API_URL}api/events/${id}?from=${from}&to=${to}`);

export const getAllLeavesAPI = ({ id }: IDetailsAPIReq) =>
  fetchWithToken(`${API_URL}api/all-leaves/${id}`);

export const applyLeaveAPI = (newLeave: INewLeave) =>
  fetchWithToken(`${API_URL}api/apply-leave`, {
    method: 'POST',
    body: JSON.stringify(newLeave)
  });

export const addAttendanceAPI = (newAttendance: INewAttendance) =>
  fetchWithToken(`${API_URL}api/attendance`, {
    method: 'POST',
    body: JSON.stringify(newAttendance)
  });

export const loginAPI = (userInfo: ILoginData) =>
  fetchWithToken(`${API_URL}login`, {
    method: 'POST',
    body: JSON.stringify(userInfo)
  });

export const getDesignationListAPI = () =>
  fetchWithToken(`${API_URL}designation`, { headers: headers });

export const registerAPI = (newUser: IRegisterAPI) =>
  fetchWithToken(`${API_URL}user`, {
    method: 'POST',
    body: JSON.stringify(newUser)
  });
