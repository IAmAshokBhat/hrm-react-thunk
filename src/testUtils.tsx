import { render, RenderOptions, waitFor } from '@testing-library/react';
import { SnackbarProvider } from 'notistack';
import { ReactElement } from 'react';
import { Provider } from 'react-redux';
import { BrowserRouter, Route, Routes } from 'react-router-dom';
import { applyMiddleware, createStore, Store } from 'redux';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { IDefaultRoot, TAny, TAnyObject } from './constants';
import { TFAction } from './contracts';
import { reducers } from './redux/reducers';
import {
  defaultArrayState,
  defaultNullState,
  defaultState
} from './redux/reducers/__tests__/fixtures';

export const middlewares = [thunk];

export const createStoreWithMiddlewares = (
  initialState: TAnyObject = {}
): Store => {
  return createStore(reducers, initialState, applyMiddleware(...middlewares));
};
interface IRenderWithProviderOptions {
  initialStoreState?: TAnyObject;
  renderFunction?: TAny;
  options?: RenderOptions;
  route?: string;
  path?: string;
}

export const renderWithProvider = (
  ui: ReactElement,
  renderOptions?: IRenderWithProviderOptions
) => {
  const {
    initialStoreState = {},
    renderFunction = render,
    options = {},
    route = '',
    path = ''
  } = renderOptions || {};

  const appStore = createStoreWithMiddlewares(initialStoreState);

  let wrappedUi = (
    <SnackbarProvider
      anchorOrigin={{
        vertical: 'top',
        horizontal: 'right'
      }}
    >
      <Provider store={appStore}>{ui}</Provider>
    </SnackbarProvider>
  );

  if (route) {
    window.history.pushState({}, '', route);

    wrappedUi = (
      <Provider store={appStore}>
        <BrowserRouter>
          <Routes>
            <Route path={path} element={ui} />
          </Routes>
        </BrowserRouter>
      </Provider>
    );
  }

  return renderFunction(wrappedUi, options);
};

type TFApiCall = (...params: TAny) => Promise<Response>;

export interface IResponseMock {
  headers?: object;
  ok?: boolean;
  status?: number;
  statusText?: string;
  type?: ResponseType;
  url?: string;
  body?: TAny | null;
  bodyUsed?: boolean;
  blob?: () => Promise<TAny>;
}

const dummyPromise = () => new Promise((resolve) => resolve('dummy'));

export class ResponseMock {
  headers: object;

  ok: boolean;

  status: number;

  statusText: string;

  type: ResponseType;

  url: string;

  body: ReadableStreamDefaultReadResult<Uint8Array> | null;

  bodyUsed: boolean;

  blob: () => Promise<TAny>;

  constructor(props: IResponseMock = {}) {
    this.headers = props.headers || {};
    this.status = props.status || 200;
    this.ok = props.ok || this.status === 200;
    this.statusText = props.statusText || 'OK';
    this.type = props.type || 'basic';
    this.url = props.url || 'http://localhost';
    this.body = props.body;
    this.bodyUsed = props.bodyUsed || this.body !== null;
    this.blob = props.blob || dummyPromise;
  }

  json: () => Promise<TAny> = () => {
    const body = this.body || '""';

    try {
      return Promise.resolve(JSON.parse(body as string));
    } catch {
      return Promise.resolve(body);
    }
  };

  text: () => Promise<string> = () => Promise.resolve(this.statusText);
}

export const mockAPICall = (func: TFApiCall, response: IResponseMock) =>
  (func as jest.Mock).mockResolvedValue(new ResponseMock(response));

export const generateActionTests = async (
  actionName: string,
  action: TFAction,
  api: TAny,
  actionPayload: TAny = []
) => {
  const response = new ResponseMock({ ok: true, body: { data: [] } });

  const mockStore = configureStore(middlewares);

  (api as jest.Mock).mockReturnValue(Promise.resolve(response));
  const store = mockStore({});

  //TODO: Fix TS error
  // @ts-ignore
  await store.dispatch(action);

  await waitFor(() => {
    expect(store.getActions()).toEqual([
      {
        type: `${actionName}_LOADING`
      },
      {
        type: `${actionName}_SUCCESS`,
        //TODO: Accept action params from function params
        payload: actionPayload
      }
    ]);
  });
};

export const generateReducerTests = (
  reducerName: string,
  reducer: TAny,
  selector: TAny,
  actionName: string,
  successValue?: TAny,
  defaultValue?: TAny,
  expectedValue?: TAny,
  reducerLabel?: string
) => {
  describe(`for ${reducerName}`, () => {
    const reducerLbl = reducerLabel
      ? reducerLabel
      : reducerName.replace('Reducer', '');
    it(`${reducerName} reducer loading`, () => {
      reducer(undefined, {
        type: `${actionName}_LOADING`,
        payload: null
      });
      // TODO: Check why its failing
      // expect(selector(defaultState).loading).toEqual(true);
      expect(selector(defaultState).value).toEqual(
        defaultValue === null ? defaultValue : []
      );
      expect(selector(defaultState).error).toEqual(null);
    });

    it(`${reducerName} reducer success`, () => {
      const successState: IDefaultRoot = {
        ...defaultState,
        [reducerLbl]: {
          ...defaultNullState,
          value: expectedValue ? expectedValue : successValue
        }
      };
      reducer(defaultArrayState, {
        type: `${actionName}_SUCCESS`,
        payload: successValue
      });

      expect(selector(successState).error).toEqual(null);
      expect(selector(successState).loading).toEqual(false);
      expect(selector(successState).value).toEqual(
        expectedValue ? expectedValue : successValue
      );
    });

    it(`${reducerName} reducer error`, () => {
      const error = { message: '', status: 1, ok: false };
      const failedState: IDefaultRoot = {
        ...defaultState,
        [reducerLbl]: {
          ...defaultNullState,
          value: [],
          error
        }
      };

      reducer(undefined, {
        type: `${actionName}_FAILURE`,
        payload: error
      });

      expect(selector(failedState).loading).toEqual(false);
      expect(selector(failedState).value).toEqual([]);
      expect(selector(failedState).error).toEqual(error);
    });
  });
};
