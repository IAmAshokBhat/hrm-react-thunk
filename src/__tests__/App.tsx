import { screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { SnackbarProvider } from 'notistack';

import App from '../App';
import {
  defaultState,
  leaveBalanceAPIResponse
} from '../components/__tests__/fixtures';
import { TAnyObject } from '../constants';
import { mockAPICall, renderWithProvider } from '../testUtils';
import * as utils from '../utils';
import { getLeaveBalanceAPI } from '../api';

jest.mock('../api', () => ({
  getLeaveBalanceAPI: jest.fn()
}));

const leaveBalanceResponse = {
  ok: true,
  body: leaveBalanceAPIResponse
};

describe('App Component', () => {
  const generateNode = (initialStoreState: TAnyObject = defaultState) =>
    renderWithProvider(
      <SnackbarProvider>
        <App />
      </SnackbarProvider>,
      {
        route: `/dashboard`,
        path: '/dashboard',
        initialStoreState
      }
    );

  beforeEach(() => {
    mockAPICall(getLeaveBalanceAPI, leaveBalanceResponse);
  });
  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should render and match snapshot', () => {
    const { container } = generateNode();

    expect(container).toMatchSnapshot();
  });

  it('should open apply leave dialog', async () => {
    //@ts-ignore
    utils.isValidToken = jest.fn().mockReturnValue(true);
    generateNode();

    const actionButton = screen.getByRole('button', {
      name: /all actions of leave management system/i
    });

    await waitFor(() => userEvent.click(actionButton));

    await waitFor(() =>
      userEvent.click(
        screen.getByRole('menuitem', {
          name: /apply leave/i
        })
      )
    );
    expect(
      screen.getByRole('heading', {
        name: /apply leave/i
      })
    ).toBeVisible();

    userEvent.click(
      screen.getByRole('button', {
        name: /cancel/i
      })
    );
    expect(
      screen.queryByRole('heading', {
        name: /apply leave/i
      })
    ).not.toBeVisible();
  });

  it('should open add attendance dialog', async () => {
    //@ts-ignore
    utils.isValidToken = jest.fn().mockReturnValue(true);
    generateNode();

    const actionButton = screen.getByRole('button', {
      name: /all actions of leave management system/i
    });

    await waitFor(() => userEvent.click(actionButton));

    await waitFor(() =>
      userEvent.click(
        screen.getByRole('menuitem', {
          name: /add attendance/i
        })
      )
    );
    await waitFor(() =>
      expect(
        screen.getByRole('heading', {
          name: /add attendance/i
        })
      ).toBeVisible()
    );

    userEvent.click(
      screen.getByRole('button', {
        name: /cancel/i
      })
    );
    expect(
      screen.queryByRole('heading', {
        name: /add attendance/i
      })
    ).not.toBeVisible();
  });
});
