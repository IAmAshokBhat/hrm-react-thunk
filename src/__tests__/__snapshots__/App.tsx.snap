// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Component should render and match snapshot 1`] = `
<div>
  <div
    class="App"
  >
    <header
      class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation4 MuiAppBar-root MuiAppBar-colorPrimary MuiAppBar-positionStatic css-hip9hq-MuiPaper-root-MuiAppBar-root"
    >
      <div
        class="MuiContainer-root MuiContainer-maxWidthXl css-19r6kue-MuiContainer-root"
      >
        <div
          class="MuiToolbar-root MuiToolbar-regular css-1cndkqf-MuiToolbar-root"
        >
          <div
            class="MuiTypography-root MuiTypography-h6 MuiTypography-noWrap css-1083xa6-MuiTypography-root"
          >
            <a
              aria-current="page"
              class="active"
              data-testid="logo-icon"
              href="/dashboard"
              style="display: block; margin: 0px; color: white; text-decoration: underline; font-weight: 600;"
            >
              <svg
                fill="white"
                height="75px"
                width="75px"
              >
                logo1.svg
              </svg>
            </a>
          </div>
          <div
            class="MuiTypography-root MuiTypography-h6 MuiTypography-noWrap css-h7cdpm-MuiTypography-root"
          >
            LOGO
          </div>
        </div>
      </div>
    </header>
    <body
      style="margin: 20px auto; max-width: 1536px;"
    />
    <div>
      <div
        class="MuiTypography-root MuiTypography-overline css-b3fch5-MuiTypography-root"
        style="text-align: center; bottom: 0px; width: 100%;"
      >
        Â©Copyright Ashok H Bhat 2022
      </div>
    </div>
  </div>
</div>
`;
